### ISSUE TITLE
common/login 404 with microsoft

### ISSUE THREAD

**Issue by ferdinandyb** [pdobsan/mailctl#5]
First off, thanks for the software, I managed to use it to log in to both my personal and work google mail. Unfortunately I have a third one at microsoft. I'm trying to access it using this config:

```
microsoft:
  auth_endpoint: https://login.microsoftonline.com/common/oauth2/v2.0/authorize
  auth_http_method: GET
  auth_params_mode: query-string
  token_endpoint: https://login.microsoftonline.com/common/oauth2/v2.0/token
  token_http_method: GET
  token_params_mode: query-string
  redirect_uri: http://localhost:8080
  auth_scope: https://outlook.office365.com/IMAP.AccessAsUser.All https://outlook.office365.com/SMTP.Send offline_access
  client_id: X
  client_secret: Y
```

I set the clien ID from the Azure overview pane, from the line saying: Application (client) ID. I also created a secret under certificates and secrets, and tried with using both the `value` and the `secret id` just in case. Both cases lead to getting a login screen with the proper background for my organisation on `localhost:8080/start`, but after entering my password I get redirected to `localhost:8080/common/login` where I get a 404 and the following displayed on screen:

```
localWebServer: request not found ... Request {requestMethod = "POST", httpVersion = HTTP/1.1, rawPathInfo = "/common/login", rawQueryString = "", requestHeaders = [("Host","localhost:8080"),("Connection","keep-alive"),("Content-Length","1634"),("Cache-Control","max-age=0"),("sec-ch-ua","\"Chromium\";v=\"104\", \" Not A;Brand\";v=\"99\", \"Google Chrome\";v=\"104\""),("sec-ch-ua-mobile","?0"),("sec-ch-ua-platform","\"Linux\""),("Origin","http://localhost:8080"),("Upgrade-Insecure-Requests","1"),("DNT","1"),("Content-Type","application/x-www-form-urlencoded"),("User-Agent","Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.79 Safari/537.36"),("Accept","text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"),("Sec-Fetch-Site","same-origin"),("Sec-Fetch-Mode","navigate"),("Sec-Fetch-User","?1"),("Sec-Fetch-Dest","document"),("Referer","http://localhost:8080/"),("Accept-Encoding","gzip, deflate, br"),("Accept-Language","hu-HU,hu;q=0.9"),("Cookie","CkTst=G1659816542347; brcap=0; clrc={%2219211%22%3a[%22yXuYioga%22%2c%22nXGRaMn2%22]}; wlidperf=FR=L&ST=1659816558812")], isSecure = False, remoteHost = 127.0.0.1:46164, pathInfo = ["common","login"], queryString = [], requestBody = , vault = , requestBodyLength = KnownLength 1634, requestHeaderHost = Just "localhost:8080", requestHeaderRange = Nothing}
```

Versions `0.6.2` and `0.6.4` produce the same results. Am I doing something wrong here?

**Comment by pdobsan** [comment]
> First off, thanks for the software, I managed to use it to log in to both my personal and work google mail.

You are welcome. If the config for your work gmail is significantly different from what is in the `services-template.yaml` would you mind to share a template. Some people had difficulties with their institutional gmail.

>  Unfortunately I have a third one at microsoft.
> ...
> Am I doing something wrong here?

To be honest, I don't know, maybe. Maybe `mailctl` does something wrong. I don't have M$ accounts. So I am putting the burden of solving M$ related problems on the users ;-) To facilitate this I provide some configuration options and debug mode.
Try to play around with the `_http_method` and `_params_mode` parameters and use `--debug`.

**Comment by ferdinandyb** [comment]
I tried to do my company auth flow for google all over again to confirm that it works but it now returns a `400` with a page saying the request was malformed instead of it working :( It definitely worked before with the config copied straight from your sample config so I'm not sure what went wrong ...

**Comment by ferdinandyb** [comment]
About the MS login, just to confirm: the redirection to `common/login/` is not something `mailctl` should be handling, but the somehow the service configuration should be changed that the redirection is made properly?

**Comment by ferdinandyb** [comment]
I also ran the authentication process with `--debug` but the last thing it prints (other than the "Now try to fetch some email!" at the end) is the html+css+js stuff for the login page, so basically only up to the point where things still work and nothing after. Do you have any suggestions where I could start poking around? 

It would also be nice if `mailctl` would end with an error if it failed, because it currently always prints the try to fetch email line, so I started manually checking if the encrypted file was created or not :)

**Comment by pdobsan** [comment]
> It would also be nice if mailctl would end with an error if it failed, because it currently always prints the try to fetch email line, so I started manually checking if the encrypted file was created or not :)

Indeed, a more elegant error handling is already on my TODO list. It is a bit more complicated than it seems.

**Comment by ferdinandyb** [comment]
As an update for the google login, I managed to solve logging in to `username@company.com`. The problem is that if you start mailctl with a domain e.g. `mailctl authorize google username@company.com` it actually takes you to a _different_ login page than the "normal" google login page. So the trick is to go `mailctl authorize google company` and then simple pick the right account on google's login page. Since now the only usage of `company` here is for the name of the file `mailctl` will create you just need to pick something that makes sense and doesn't mess with your other accounts.

**Comment by ferdinandyb** [comment]
Nice, I actually tried the same trick now with MS, but now it tells me that an admin needs to authorize my app :) Great

**Comment by ferdinandyb** [comment]
Ok, so one step closer again, using Thunderbird's client_id and the same trick as with google (pass name of organisation as the "username" and then switching in microsoft's auth screen to the organisation login), I managed to grant access to log in and grant access to Thunderbird, but now with the debug flag I get this:

```
localWebServer:
fetchAuthRecord: Error in $: Failed reading: not a valid json value
CallStack (from HasCallStack):
  error, called at lib/MailCtl/Authorization.hs:278:17 in mailctl-0.6.4-f4efc9a3e27c412daaa05fd4c22ce22b7d21cf8b50d56ab76f3bc63a27ae8204:MailCtl.Authorization
```

**Comment by pdobsan** [comment]
> As an update for the google login, I managed to solve logging in to `username@company.com`.

That is great! I am going to incorporate your trick into the docs :+1:

**Comment by ferdinandyb** [comment]
I think he has _exactly_ the same config as me, except for him having a cron indicator configured, which shouldn't matter I guess, so I'm a bit baffled, but could be that my uni has some different settings than his. At this point I'm quite out of ideas. On the other hand, it seems currently, I can actually forgo `mailctl` as MS accepts password logins. If you have an idea on what to try I'd be happy to hammer away, but otherwise I think I'll close the issue.

**Comment by pdobsan** [comment]
> I think he has _exactly_ the same config as me, except for him having a cron indicator configured, which shouldn't matter I guess, so I'm a bit baffled, but could be that my uni has some different settings than his.

You might want to contact him to _compare notes_ so to speak.

(Cron doesn't matter and I am going to remove the coupling anyway.)

**Comment by hsseung** [comment]
I'm using v0.6.5 and have exactly the same issue as https://github.com/pdobsan/mailctl/issues/5#issue-1330840646 when I try `mailctl authorize microsoft username@school.edu`.
Namely, Microsoft gives me the SSO prompt from my school, but when I enter the password, I get redirected to localhost:8080/common/login with an error message in the browser.

I also tried `mailctl authorize microsoft school.edu`. Then Microsoft displays the SSO for a split second, and redirects to `login.live.com` with the error message "We're unable to complete your request
unauthorized_client: The client does not exist or is not enabled for consumers. If you are the application developer, configure a new application through the App Registrations in the Azure Portal at https://go.microsoft.com/fwlink/?linkid=2083908."

**Comment by ferdinandyb** [comment]
try `mailctl authorize microsoft school` so it is not interpreted as a
domain


hsseung ***@***.***> ezt írta (időpont: 2022. aug. 15., Hét
1:31):

**Comment by guille-c89** [comment]
I think I have a clue here, some tenants have the "Keep me signed in (KMSI)" promtp activated: <https://docs.microsoft.com/en-us/azure/active-directory/fundamentals/keep-me-signed-in>

So I wonder if, similar to #3, there is unhandled redirection.

**Comment by guille-c89** [comment]
Testing with the `mutt_oauth.py` script[1], I found that my tenant doesn't accept Thunderbird as an email client. In this scenario, I think I'll ask to tenant's admin for a valid `client_id` and `client_secret`.


References:
1. <https://gitlab.com/muttmua/mutt/-/blob/master/contrib/mutt_oauth2.py>

**Comment by ferdinandyb** [comment]
@guille-c89 that's very strange, did you check if you can use Thunderbird itself to connect? I'd find it strange if you were not able to.

**Comment by pdobsan** [comment]
I believe the authorization problem with Microsoft has been solved with release 0.7.3.
Update and modify the `microsoft` entry in your `services.yaml` file according to the new template.
